File: src/audio/tagging/chapter/Chapter.java
package audio.tagging.chapter;

import java.util.Objects;
import javafx.util.Duration;
import util.units.Dur;

/**
 * Chapter is a text associated with specific point of time in a song item.
 * <p/>
 * Chapter always has time and text assigned to it. Time must always be specified
 * while text is "" by default.
 */
public final class Chapter implements Comparable<Chapter> {

	private Dur time;
	private String info = "";

	public Chapter(Duration time) {
		this(time, "");
	}

	public Chapter(Duration time, String _info) {
		setTime(time);
		info = _info;
	}

	/**
	 * Creates new chapter by parsing the string.
	 *
	 * @param text string to parse. Should be a result of {@link #toString()}.
	 * @throws IllegalArgumentException if not parsable String.
	 */
	public Chapter(String text) {
		int i = text.indexOf('-');
		if (i==-1) throw new IllegalArgumentException("Not parsable chapter string: " + text);
		String s = text.substring(0, i);
		try {
			setTimeInMillis(Double.parseDouble(s));
		} catch (NumberFormatException e) {
			throw new IllegalArgumentException("Not parsable chapter string: " + text);
		}
		info = text.substring(i + 1, text.length());
	}

	/**
	 * Gets time as Duration with granularity of 1 millisecond.
	 */
	public Dur getTime() {
		// time must always be equivalent to mathematical integer
		assert time.toMillis()==Math.rint(time.toMillis());

		return time;
	}

	/**
	 * Sets time. Time denotes the position within the song this chapter is
	 * associated with. It is also a unique identifier and two chapters with the
	 * same time should not exist. They will be considered equal.
	 */
	public void setTime(Duration time) {
		setTimeInMillis(time.toMillis());
	}

	private void setTimeInMillis(double millis) {
		// round to decimal number before assigning
		this.time = new Dur(Math.rint(millis));
	}

	/** Returns the text. Never null. Default value is "". */
	public String getText() {
		return info;
	}

	/** Sets text. */
	public void setText(String info) {
		this.info = info;
	}

	/**
	 * @return true if and only if both chapters and their time is the same.
	 */
	@Override
	public boolean equals(Object o) {
		return this==o || (o instanceof Chapter && getTime().equals(((Chapter) o).getTime()));
	}

	@Override
	public int hashCode() {
		int hash = 3;
		hash = 19*hash + Objects.hashCode(this.time);
		return hash;
	}

	/**
	 * Compares chapter by natural order - by time.
	 */
	@Override
	public int compareTo(Chapter o) {
		return time.compareTo(o.time);
	}

	/**
	 * Use to convert to String, which can be then reconstructed back to Chapter
	 * with {@link #Chapter(java.lang.String)}.
	 * Example: "9800-New Chapter".
	 */
	@Override
	public String toString() {
		return Math.rint(time.toMillis()) + "-" + info;
	}
}

File: src/audio/tagging/chapter/Chapter.kt
package audio.tagging.chapter

import javafx.util.Duration
import util.text.Strings
import util.units.Dur
import java.util.*

/** Chapter is a text associated with specific point of time in a song. */
class Chapter: Comparable<Chapter> {

    /** Time of this chapter from the beginning of the song. Always equivalent to mathematical integer */
    var time: Dur

    /** Text of this chapter. */
    var text = ""

    @JvmOverloads constructor(time: Duration, text: String = "") {
        this.time = time.toMillis().computeTimeFromMs()
        this.text = text
    }

    /**
     * @param text string to parse, should be a result of [toString]
     * @throws RuntimeException if not parsable String
     */
    constructor(text: String) {
        val i = text.indexOf(separator)
        if (i==-1) throw IllegalArgumentException("Not parsable chapter string: $text")     // TODO: handle better
        this.time = text.substring(0, i).toDoubleOrNull()?.computeTimeFromMs() ?: Dur(0.0)
        this.text = text.substring(i+1, text.length)
    }

    /** @return true iff both chapters text and their time */
    override fun equals(other: Any?) = this===other || other is Chapter && time==other.time

    override fun hashCode() = 19*3+Objects.hashCode(this.time)

    /** @return result of comparison by time */
    override fun compareTo(other: Chapter) = time.compareTo(other.time)

    /** @return text, which can be reconstructed back to Chapter with [Chapter], e.g.: "9800-New Chapter" */
    override fun toString() = "${time.toMillis()}$separator$text"

}

class Chapters: ArrayList<Chapter>(), Strings {
    override val strings: Sequence<String> get() = asSequence().map { it.text }
}

private fun Double.computeTimeFromMs() = Dur(Math.rint(this)) // round to decimal number before assigning

private val separator = "-"